### DISCUSSION TITLE
STM32 Webserver combined to I2S. No free descriptors problem.

### DISCUSSION THREAD

**Discussion by PietroElectr** [cesanta/mongoose#2828] [Ask questions, get answers] [answered]
Good morning,

I'm working on a project able to manage an audio stream as input using an external 24bit ADC set at 44100Hz and sending the sampled packetized data using mongoose webserver (websocket).
The EVM that I'm using is a STM32H723ZG Nucleo.
I decided to write two different firmwares
1) Collecting data from ADC using I2S triggering some callback functions. In order to manage the input audio stream I'm forced to use the I2S protocol (constraint given by the ADC) and since it is a big stream I Thought to use the DMA to reduce the load on the CPU.

2) Following your guide connecting the STM32H7 to the wifi through ethernet cable and to access to the html/js files. So i can communicate with the board, asking and sending data.
I did it and they both work separately!

Then starting from the Mongoose firmware i tried to integrate the part related to I2S.

![image](https://github.com/cesanta/mongoose/assets/103196253/cd47234f-327e-4b6c-95fe-579d02e62233)
![image](https://github.com/cesanta/mongoose/assets/103196253/1887e916-3cfb-4f0f-9b8d-e5c8a9886baa)

The main function is the following one (without managing the audio stream, just launching the I2S_DMA)

int main(void)
{

  /* USER CODE BEGIN 1 */

  /* USER CODE END 1 */

  /* MPU Configuration--------------------------------------------------------*/
  MPU_Config();

  /* MCU Configuration--------------------------------------------------------*/

  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
  HAL_Init();

  /* USER CODE BEGIN Init */

  /* USER CODE END Init */

  /* Configure the system clock */
  SystemClock_Config();

  /* USER CODE BEGIN SysInit */

  /* USER CODE END SysInit */

  /* Initialize all configured peripherals */
  MX_GPIO_Init();
  MX_DMA_Init();
  MX_USART3_UART_Init();
  MX_ETH_Init();
  MX_I2C1_Init();
  MX_I2S3_Init();
  /* USER CODE BEGIN 2 */

  /* USER CODE END 2 */

  /* Infinite loop */
  /* USER CODE BEGIN WHILE */

  mg_mgr_init(&mgr);
  mg_log_set(MG_LL_DEBUG);

  // On STM32Fxx, use _stm32f suffix instead of _stm32h
  struct mg_tcpip_driver_stm32h_data driver_data = {.mdc_cr = 4};
  struct mg_tcpip_if mif = {.mac = {2, 3, 4, 5, 6, 7},
                            // Uncomment below for static configuration:
                            // .ip = mg_htonl(MG_U32(192, 168, 0, 223)),
                            // .mask = mg_htonl(MG_U32(255, 255, 255, 0)),
                            // .gw = mg_htonl(MG_U32(192, 168, 0, 1)),
                            .driver = &mg_tcpip_driver_stm32h,
                            .driver_data = &driver_data};
  NVIC_EnableIRQ(ETH_IRQn);
  mg_tcpip_init(&mgr, &mif);
  mg_http_listen(&mgr, "http://0.0.0.0:80", fn, NULL);
  HAL_I2S_Receive_DMA(&hi2s3, (uint16_t*)samples, sizeof(samples));

  while (1) {

    mg_mgr_poll(&mgr, 0);
   
    /* USER CODE END WHILE */

    /* USER CODE BEGIN 3 */
  }
  /* USER CODE END 3 */
}

The two critical functions are of course 
HAL_I2S_Receive_DMA(&hi2s3, (uint16_t*)samples, sizeof(samples));
and
mg_mgr_poll(&mgr, 0);

that gives to me the following output:

![image](https://github.com/cesanta/mongoose/assets/103196253/8211ead6-5ebd-41bc-a5e6-8e1d74bc5efe)

Now, I'm new and I'm trying to figure out what is happening.
I tried to change the I2S module since the STM32H7 has 4 different ports. I also tried to change the channels but in both cases
I still had the problem. So I would kindly to ask if you have any hints or proposed solution in order to solve the "descriptor problem" or any suggestion to reach the goal, so managing the audio stream sending it through Mongoose.

If you need any information don't hesitate to ask.
Thank you for your time.

pb

**Comment by scaprile** [comment, marked as answer]
Mongoose does not know, nor cares, about I2S or your hardware or your application.
Whatever, you do, you break something, change something, modify something.
Your problem is probably the MAC DMA not being able to access descriptor memory. See your reference manual, see where your memory is allocated, and leave those memory areas where the Ethernet Controller can read/write, available for it.
